pandas/core/internals/managers.py

```diff
@@ -1952,8 +1952,25 @@ def _compare_or_regex_search(a, b, regex=False):
         # GH#29553 avoid deprecation warnings from numpy
         result = False
     else:
+        # GH#32621 use mask to avoid comparing to NAs
+        if is_a_array and not is_b_array:
+            mask = np.reshape(~(isna(a)), a.shape)
+        elif is_b_array and not is_a_array:
+            mask = np.reshape(~(isna(b)), b.shape)
+        elif is_a_array and is_b_array:
+            mask = ~(isna(a) | isna(b))
+
+        if is_a_array:
+            a = a[mask]
+        if is_b_array:
+            b = b[mask]
         result = op(a)
 
+        if isinstance(result, np.ndarray):
+            tmp = np.zeros(mask.shape, dtype=np.bool)
+            tmp[mask] = result
+            result = tmp
```

Comment:
why is this not just
```result[mask] = False```

Reply:
Because the shape of the mask can differ to that of the result, since we may select to compare only a subset of a's or b's elements (i.e. some of them may be NAs).

Reply:
ok can you add a comment on what this is doing (simnilar to  what you just said)